pkg/dispers/dispers/metadata.go

```diff
@@ -0,0 +1,141 @@
+)
+
+// Only the Conductor should save metadata
+var prefix = prefixer.ConductorPrefixer
+
+// Metadata are written on the confuctor's database. The querier can read those Metadata to know his query's state
+type Metadata struct {
+	Start       time.Time `json:"start,omitempty"`
+	End         time.Time `json:"end,omitempty"`
+	Duration    string    `json:"duration,omitempty"`
+	Host        string    `json:"host,omitempty"`
+	Tags        []string  `json:"tags,omitempty"`
+	Outcome     bool      `json:"outcome,omitempty"`
+	Error       string    `json:"error,omitempty"`
+	Name        string    `json:"name,omitempty"`
+	Description string    `json:"description,omitempty"`
+	Output      string    `json:"output,omitempty"`
+}
+
+// NewMetadata returns a new Metadata object
+func NewMetadata(name string, description string, tags []string) Metadata {
+	now := time.Now()
+	host, _ := os.Hostname()
+
+	return Metadata{
+		Start:       now,
+		Description: description,
+		Name:        name,
+		Tags:        tags,
+		Host:        host,
+	}
+}
+
+// Close finish writting Metadata
+func (m *Metadata) Close(msg string, err error) error {
+	m.End = time.Now()
+	m.Duration = (m.End.Sub(m.Start)).String()
+	m.Outcome = (err == nil)
+	m.Output = msg
+	if err != nil {
+		m.Error = err.Error()
+	}
+	return nil
+}
+
+type MetadataDoc struct {
+	MetaID   string     `json:"_id,omitempty"`
+	MetaRev  string     `json:"_rev,omitempty"`
+	QueryID  string     `json:"query,omitempty"`
+	Metadata []Metadata `json:"metadata,omitempty"`
+}
+
+func (t *MetadataDoc) ID() string {
+	return t.MetaID
+}
+
+func (t *MetadataDoc) Rev() string {
+	return t.MetaRev
+}
+
+func (t *MetadataDoc) DocType() string {
+	return "io.cozy.metadata"
+}
+
+func (t *MetadataDoc) Clone() couchdb.Doc {
+	cloned := *t
+	return &cloned
+}
+
+func (t *MetadataDoc) SetID(id string) {
+	t.MetaID = id
+}
+
+func (t *MetadataDoc) SetRev(rev string) {
+	t.MetaRev = rev
+}
+
+// RetrieveMetadata get Metadata from a MetadataDoc in CouchDB
+func RetrieveMetadata(queryid string) ([]MetadataDoc, error) {
+
+	couchdb.EnsureDBExist(prefix, "io.cozy.metadata")
+
+	// Check if doc existing
+	var out []MetadataDoc
+	req := &couchdb.FindRequest{Selector: mango.Equal("query", queryid)}
+	err := couchdb.FindDocs(prefix, "io.cozy.metadata", req, &out)
+	if err != nil {
+		return []MetadataDoc{}, err
+	}
+	return out, nil
+}
+
+// Push adds metadata to the MetadataDoc in Conductor's database
+func (m *Metadata) Push(queryid string) error {
+
+	out, err := RetrieveMetadata(queryid)
+	if err != nil {
+		return err
+	}
+
+	if len(out) == 1 {
+		// Retrieve doc
+		doc := out[0]
+		// Update doc
+		newArrayMetadata := make([]Metadata, len(doc.Metadata)+1)
+		for index, item := range doc.Metadata {
+			newArrayMetadata[index] = item
+		}
+		newArrayMetadata[len(doc.Metadata)] = *m
+		doc.Metadata = newArrayMetadata
+		// Upload doc
+		err := couchdb.UpdateDoc(prefix, &doc)
+		if err != nil {
+			return err
+		}
```

Comment:
If I understand, each new update creates a new entry in the Metadata array? So this is more an append than a update? Why not a real document update with the new metadata erasing the previous ones?

Reply:
> If I understand, each new update creates a new entry in the Metadata array? So this is more an append than a update?

Yes, this is it ! I think I did not know "append" exist in Go when I build this. I'm going to change that. 

> Why not a real document update with the new metadata erasing the previous ones?

Not sure to understand. May be we should speak about this from Zoom. I planned several times to rename metadata and call it "logs" because It may be a proper name. I think we should discuss together to be sure of what we want Metadata/Logs to do. 

I wanted metadata to give information about every single step of the query. I wanted it to be a way to pass error message. But regarding another comment you made today, I think my dealing of errors that occured in another server is very bad for the moment.

Why erasing the previous metadata if we want metadata to tell us, for example : 
- How long did it take to discuss with each stack ?  

A month ago, I wanted to use metadata (I should totally have named that logs, I don't know why I did not change that) to create a dashboard to follow the process.

Reply:
Ok I understand better what you want to do. Let's discuss about it tomorrow!